net: fix banning and disallow sending messages before receiving verack

theuni opened the pull request:
This is the last of my net issues for 0.14. As discussed with @TheBlueMatt and @gmaxwell.

Fixes for a few problems discovered while running a network stress/fuzzer:
- Remote nodes weren't always banned when they hadn't yet sent a verack. Regression from 7a8c2519015650acd51eaf42719f04e53f839bbe.
- Require a verack before sending any non-handshake messages. This is much more straightforward behavior, and allows for tests to be easily written
- Now that there's a sane model for testing, add checks for leaky messages sent out before the handshake is complete, as well as for banning in those cases.

theuni committed:
net: correctly ban before the handshake is complete

7a8c251901 made a change to avoid getting into SendMessages() until the
version handshake (VERSION + VERACK) is complete. That was done to avoid
leaking out messages to nodes who could connect, but never bothered sending
us their version/verack.

Unfortunately, the ban tally and possible disconnect are done as part of
SendMessages(). So after 7a8c251901, if a peer managed to do something
bannable before completing the handshake (say send 100 non-version messages
before their version), they wouldn't actually end up getting
disconnected/banned. That's fixed here by checking the banscore as part of
ProcessMessages() in addition to SendMessages().

TheBlueMatt committed:
qa: mininode learns when a socket connects, not its first action

TheBlueMatt committed:
qa: Expose on-connection to mininode listeners

theuni committed:
net: parse reject earlier

Prior to this change, all messages were ignored until a VERSION message was
received, as well as possibly incurring a ban score.

Since REJECT messages can be sent at any time (including as a response to a bad
VERSION message), make sure to always parse them.

Moving this parsing up keeps it from being caught in the
if (pfrom->nVersion == 0) check below.

theuni committed:
net: require a verack before responding to anything else

7a8c251901 made this logic hard to follow. After that change, messages would
not be sent to a peer via SendMessages() before the handshake was complete, but
messages could still be sent as a response to an incoming message.

For example, if a peer had not yet sent a verack, we wouldn't notify it about
new blocks, but we would respond to a PING with a PONG.

This change makes the behavior straightforward: until we've received a verack,
never send any message other than version/verack/reject.

The behavior until a VERACK is received has always been undefined, this change
just tightens our policy.

This also makes testing much easier, because we can now connect but not send
version/verack, and anything sent to us is an error.

theuni committed:
qa: add a test to detect leaky p2p messages

This is certainly not exhaustive, but it's better than nothing. Adds checks
for:

- Any message received before sending a version
- Any message received other than version/reject before sending a verack

It also tries to goad the remote into sending a pong, address, or block
announcement.

theuni commented:
I see that @TheBlueMatt and I managed to make almost the exact same change:
2a278cdaf6ea46dc85a61a37a12a8acd7acd5670 vs a5032b5b0c55c90a8e4df658d85d99824cf4699d. I'm happy to rebase and drop mine if his goes in first.

rebroad commented:
What is the risk by not merging this?

TheBlueMatt commented:
Concept ACK, will give this a more full review after breakfast.

On February 8, 2017 1:46:00 AM EST, Cory Fields <notifications@github.com> wrote:
>This is the last of my net issues for 0.14. As discussed with
>@TheBlueMatt and @gmaxwell.
>
>Fixes for a few problems discovered while running a network
>stress/fuzzer:
>- Remote nodes weren't always banned when they hadn't yet sent a
>verack. Regression from 7a8c2519015650acd51eaf42719f04e53f839bbe.
>- Require a verack before sending any non-handshake messages. This is
>much more straightforward behavior, and allows for tests to be easily
>written
>- Now that there's a sane model for testing, add checks for leaky
>messages sent out before the handshake is complete, as well as for
>banning in those cases.
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/bitcoin/bitcoin/pull/9720
>
>-- Commit Summary --
>
>  * net: correctly ban before the handshake is complete
>  * net: parse reject earlier
>  * net: require a verack before responding to anything else
>  * qa: allow for a node that does not send an initial version message
>  * qa: add a test to detect leaky p2p messages
>
>-- File Changes --
>
>    M qa/pull-tester/rpc-tests.py (1)
>    A qa/rpc-tests/p2p-leaktests.py (135)
>    M qa/rpc-tests/test_framework/mininode.py (21)
>    M src/net_processing.cpp (115)
>
>-- Patch Links --
>
>https://github.com/bitcoin/bitcoin/pull/9720.patch
>https://github.com/bitcoin/bitcoin/pull/9720.diff
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/9720

instagibbs reviewed (approved):
utACK 50ce6755e024bc849ef8681f9aaa81ff3f3bca57

I'd also like the commit referred to in 0abd2ccfaa24fdd1e5d24271e2360c4d9d64b662 's commit message to include its commit hash for easier understanding.

theuni commented:
I got a few questions about these changes, so I've updated the commit messages to provide (I hope) better back-story/context.

theuni commented on src/net_processing.cpp:
```diff
@@ -2594,6 +2594,35 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
     return true;
 }
 
+static bool SendRejectsAndCheckBan(CNode* pnode, CConnman& connman)
```
Heh, i waffled back and forth on this. Sure.

TheBlueMatt replied:
The usual fix for waffling is comment and clarify name :p

theuni commented on src/net_processing.cpp:
```diff
@@ -2594,6 +2594,35 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
+    AssertLockHeld(cs_main);
+    CNodeState &state = *State(pnode->GetId());
+
+    BOOST_FOREACH(const CBlockReject& reject, state.rejects)
```
Will do.

kallewoof replied:
Nit: avoid `BOOST_FOREACH` if possible (though this was moved code, so perhaps not applicable, but..).
```C++
for (const CBlockReject& reject : state.rejects) [...]
```

laanwj replied:
No, for moved code this is not applicable. To avoid confusion, let's keep code style changes, moves and bugfixes separate where possible.

kallewoof reviewed (commented):
utACK, did not review tests

TheBlueMatt commented on src/net_processing.cpp:
```diff
@@ -2706,8 +2735,11 @@ bool ProcessMessages(CNode* pfrom, CConnman& connman, const std::atomic<bool>& i
+        if (!fRet) {
             LogPrintf("%s(%s, %u bytes) FAILED peer=%d\n", __func__, SanitizeString(strCommand), nMessageSize, pfrom->id);
+            LOCK(cs_main);
+            SendRejectsAndCheckBan(pfrom, connman);
```
There are a few cases it looks like we return true after setting misbehaving, so this probably needs to be checked either way (or those cases need updating - but that might result in double-logging).

theuni replied:
I was trying to avoid the cs_main lock for each message received, but you're right that it means that we rely on a return value that's not very well-defined.

I'll just make it unconditional. Until we have parallel processing, there's not much harm in that.

TheBlueMatt replied:
Sounds good.

TheBlueMatt commented:
utACK df1a32392933a4f716c53d62703a56e8d8bda9da, did not review tests.

gmaxwell reviewed (approved):
ACK (only tested lightly)

jnewbery commented on qa/rpc-tests/p2p-leaktests.py:
```diff
@@ -0,0 +1,135 @@
+from test_framework.mininode import *
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import *
+import logging
```
You can remove this. logging isn't being used.

theuni replied:
ok

jnewbery commented on qa/rpc-tests/p2p-leaktests.py:
```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+# Copyright (c) 2017 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
```
Please add a docstring describing what this test case is doing.

theuni replied:
ok

jnewbery commented on qa/rpc-tests/p2p-leaktests.py:
```diff
@@ -0,0 +1,135 @@
+    def check(self):
+        def done():
+            return self.done
+        return wait_until(done, timeout=10) and not self.unrequested_msg
```
I'm not entirely sure what this is achieving. I think you can remove `check()` and `self.done` entirely.

You can check directly that no unexpected messages have been received at the end of the test (see my comment below).

jnewbery commented on qa/rpc-tests/p2p-leaktests.py:
```diff
@@ -0,0 +1,135 @@
+        self.connection.send_message(message)
+
+    def bad_message(self, message):
+        self.unrequested_msg = True
```
Is `unexpected_message` a better name than `unrequested_message`?

theuni replied:
seems about the same to me, but sure

jnewbery commented on qa/rpc-tests/p2p-leaktests.py:
```diff
@@ -0,0 +1,135 @@
+
+    def bad_message(self, message):
+        self.unrequested_msg = True
+        raise Exception("should not have received message: %s" % message.command)
```
I don't think you should be raising an exception in the network thread. Just set `unexpected_message` to True (and optionally print some debug info)

theuni replied:
ok

jnewbery commented on qa/rpc-tests/p2p-leaktests.py:
```diff
@@ -0,0 +1,135 @@
+    def on_sendcmpct(self, conn, message): self.bad_message(message)
+    def on_cmpctblock(self, conn, message): self.bad_message(message)
+    def on_getblocktxn(self, conn, message): self.bad_message(message)
+    def on_blocktxn(self, conn, message): self.bad_message(message)
```
Sigh. This is verbose and will be incomplete every time new p2p message types are added.

I was tempted to say just override the `deliver()` message from `NodeConnCB` but that just feels ugly (`deliver()` should really be a private method and the testcases just override use the on_ callbacks).

I can't think of anything better here.

theuni replied:
yes, i hate this too.

Too much to do here but for the future: how about a base class for messages. Then there's an overridable dispatcher that forwards to the individual callbacks. So here, I would just override the dispatcher and check the message type.

Basically just another thin layer under deliver()

jnewbery commented on qa/rpc-tests/p2p-leaktests.py:
```diff
@@ -0,0 +1,135 @@
+    def on_reject(self, conn, message): pass
+    def on_verack(self, conn, message): pass
+
+    # When version is received, don't replay with a verack. Instead, see if the
```
s/replay/reply

theuni replied:
ok

jnewbery commented on qa/rpc-tests/p2p-leaktests.py:
```diff
@@ -0,0 +1,135 @@
+    def add_options(self, parser):
+        parser.add_option("--testbinary", dest="testbinary",
+                          default=os.getenv("BITCOIND", "bitcoind"),
+                          help="Binary to test max block requests behavior")
```
I don't think this parser.add_option line is required.

theuni replied:
ok. Definitely the c/p from the wrong test isn't needed :)

jnewbery commented on qa/rpc-tests/p2p-leaktests.py:
```diff
@@ -0,0 +1,135 @@
+    def __init__(self):
+        super().__init__()
+        self.num_nodes = 1
+        self.banscore = 10
```
Since banscore isn't changing I'd prefer to make it a global constant.

theuni replied:
I think i'd rather keep this local so that we can use different scores for different tests.

jnewbery commented on qa/rpc-tests/p2p-leaktests.py:
```diff
@@ -0,0 +1,135 @@
+    def setup_network(self):
+        extra_args = [['-debug', '-banscore='+str(self.banscore)]
+                      for i in range(self.num_nodes)]
+        self.nodes = start_nodes(self.num_nodes, self.options.tmpdir, extra_args)
```
There's only one node here. You can replace the above three lines with:

```python
self.nodes = [start_node(0, self.options.tmpdir, ['-debug', '-banscore='+str(self.banscore)])]
```

theuni replied:
ok

MarcoFalke replied:
Please leave it as is for now. My goal was to generalize this logic and get rid of most `setup_network` methods in test as the only thing they commonly do is fire up the requested number of nodes and nothing else. So there would be no need to overwrite this function if you want a simple network set up.

MarcoFalke replied:
Also, it would be easier to adapt if the number of nodes changes.

jnewbery commented on qa/rpc-tests/p2p-leaktests.py:
```diff
@@ -0,0 +1,135 @@
+        no_verack_idlenode.add_connection(connections[2])
+
+        NetworkThread().start()  # Start up network handling in another thread
+        self.nodes[0].generate(1)
```
not needed?

theuni replied:
Very needed, but needs a comment. We generate a block, and wait a few secs to make sure that it's not relayed to the nodes who haven't versioned/veracked yet.

jnewbery commented on qa/rpc-tests/p2p-leaktests.py:
```diff
@@ -0,0 +1,135 @@
+
+        assert(no_version_bannode.check())
+        assert(no_version_idlenode.check())
+        assert(no_verack_idlenode.check())
```
Might as well just check the variable directly:

```python
        assert(not no_version_bannode.unexpected_message)
        assert(not no_version_idlenode.unexpected_message)
        assert(not no_verack_idlenode.unexpected_message)
```

theuni replied:
ok

jnewbery commented on qa/rpc-tests/p2p-leaktests.py:
```diff
@@ -0,0 +1,135 @@
+        assert(no_version_bannode.check())
+        assert(no_version_idlenode.check())
+        assert(no_verack_idlenode.check())
+
```
I recommend you disconnect all the connections to the node at the end of the test:

```python
        # Disconnect all peers
        [conn.disconnect_node() for conn in connections]
```

theuni replied:
ok

jnewbery reviewed (commented):
Test looks good @theuni ! I've added a bunch of style comments.

theuni commented:
@jnewbery Thanks for the great test review. I'm not sure I'll have time to get to this before Sunday, so let's not let it hold back merge if a few more ACKs come in. I'll for sure fix up the tests post-merge if that's the case.

theuni commented:
Fixed up the tests and squashed. Only the tests changed, the bitcoin code is exactly the same as before squash. I've archived the old branch here: https://github.com/theuni/bitcoin/commits/fix-ban2 in case anyone wants to compare.

I replaced my mininode changes with @TheBlueMatt's commits from #9715, so that these won't conflict with eachother.
